### PR TITLE
Remove WriterT from the transformer stack #13

### PR DESCRIPTION
**Pull request by jhenligne**

### CODE REVIEW

File: src/ConfCrypt/Commands.hs [hunk 1]
```diff
@@ -52,11 +54,11 @@ data FileAction
 --
 -- In reality the return type of 'evalutate' is 'Text', this needs to be cleaned up in the upcoming version.
 class Monad m => Command a m where
-    evaluate :: a -> m ()
+    evaluate :: a -> m [T.Text]
 
 -- | Read and return the full contents of an encrypted file. Provides support for using a local RSA key or an externl KMS service
 data ReadConfCrypt = ReadConfCrypt
-instance (Monad m, MonadDecrypt (ConfCryptM m key) key) => Command ReadConfCrypt (ConfCryptM m key) where
+instance (MonadIO m, MonadDecrypt (ConfCryptM m key) key) => Command ReadConfCrypt (ConfCryptM m key) where
```
[the remark this answers is not in the document]
Comment (jhenligne) [reply src/ConfCrypt/Commands.hs]: Initially saving changes to file was done in `writeFullContentsToBuffer` to mimic the current behavior and `MonadIO` was required to call this function.
Then I noticed writing to a file was not needed in all cases so I moved this code to Engine.hs but the constraint on `MonadIO` remained.
It has been reverted to `Monad` in the last commit.

File: app/ConfCrypt/CLI/Engine.hs [hunk 2]
```diff
@@ -92,14 +95,23 @@ run parsedArguments = do
         injectPrivateKey :: PrivateKey -> (ConfCryptFile, ()) -> (ConfCryptFile, TextKey PrivateKey)
         injectPrivateKey key (conf, _) = (conf, TextKey key)
 
+        hasToBeSaved (AC _ _) = True
```
Comment (ChrisCoffey) [inline app/ConfCrypt/CLI/Engine.hs]: This is a nice addition, but it changes the semantics of the program quite a bit and belongs in a separate PR. In order to integreate with other tools more easily, `confcrypt` actually prints its output to stdout rather than performing inplace updates of the config file.

That being said, it would be  great to separately add this function with an `inplace` command line flag to let users select this behavior.

File: src/ConfCrypt/Commands.hs [hunk 3]
```diff
@@ -79,9 +81,9 @@ instance (Monad m, MonadDecrypt (ConfCryptM m key) key) => Command GetConfCrypt
     evaluate (GetConfCrypt name) = do
         (ccFile, ctx) <- ask
         let mParam = find ((==name) . paramName) (parameters ccFile)
-        traverse_ (decrypt ctx) mParam
+        traverse (decrypt ctx) $ maybeToList mParam
         where
-            decrypt ctx = tell . pure <=< decryptValue ctx . paramValue
+            decrypt ctx = pure <=< decryptValue ctx . paramValue
```
Comment (ChrisCoffey) [inline src/ConfCrypt/Commands.hs]: Why does this use `pure <=<`? that's equivalent to the rhs of `<=<`

File: src/ConfCrypt/Validation.hs [hunk 4]
```diff
@@ -52,29 +52,29 @@ parameterTypesMatchSchema key ConfCryptFile {parameters} =
     where
         decryptAndCompare Parameter {paramName, paramValue, paramType} =
             catchError (runRule paramType paramName =<< decryptValue key paramValue)
-                       (const $ tell ["Error: Could not decrypt " <> paramName])
+                       (pure $ liftIO $ putStrLn ("Error: Could not decrypt " <> T.unpack paramName))
```
Comment (ChrisCoffey) [inline src/ConfCrypt/Validation.hs]: Rather than `putStrLn`, this should accumulate errors. By adding the `MonadIO` constraint and using `putStrLn`, it becomes much more difficult to test this function.

File: src/ConfCrypt/Validation.hs [hunk 5]
```diff
@@ -52,29 +52,29 @@ parameterTypesMatchSchema key ConfCryptFile {parameters} =
...
         runRule paramType paramName val =
             case paramType of
                 Nothing -> pure ()
                 Just CInt | all isDigit $ T.unpack val -> pure ()
                 Just CBoolean | T.toLower val == "true" || T.toLower val == "false" -> pure ()
                 Just CString | not (T.null val) -> pure ()
-                Just CString | T.null val -> tell ["Warning: "<> paramName <> " is empty"]
-                Just pt -> tell ["Error: "<> paramName <> " does not match the schema type " <> typeToOutputString pt]
+                Just CString | T.null val -> liftIO $ putStrLn ("Warning: "<> T.unpack paramName <> " is empty")
```
Comment (ChrisCoffey) [inline src/ConfCrypt/Validation.hs]: Same here with accumulating the `Text` warning messages rather than printing them.

File: src/ConfCrypt/Validation.hs [hunk 6]
```diff
@@ -52,29 +52,29 @@ parameterTypesMatchSchema key ConfCryptFile {parameters} =
...
+                Just pt -> liftIO $ putStrLn ("Error: "<> T.unpack paramName <> " does not match the schema type " <> T.unpack (typeToOutputString pt))
 
 -- | Raise an error if there are parameters without a schema
-logMissingSchemas :: (Monad m, MonadWriter [T.Text] m) =>
+logMissingSchemas :: MonadIO m =>
```
Comment (ChrisCoffey) [inline src/ConfCrypt/Validation.hs]: All of the `MonadIO` constraints can be removed once `Text` messages are accumulated.

File: test/ConfCrypt/Commands/Tests.hs [hunk 7]
```diff
@@ -59,7 +60,7 @@ modifyFileProperties = testGroup "modify file properties" [
 
     ]
 
-bufferWriteProperties :: TestTree
+{-bufferWriteProperties :: TestTree
```
Comment (ChrisCoffey) [inline test/ConfCrypt/Commands/Tests.hs]: In theory this should still be testable using the exact same logic, minus the `execWriter` call. Please restore the test.

File: test/ConfCrypt/Commands/Tests.hs [hunk 8]
```diff
@@ -68,12 +69,14 @@ bufferWriteProperties  = testGroup "Buffer write" [
...
                   parseRes
-    ]
+    ]-}
 
 readTests :: TestTree
 readTests = testGroup "Read" [
     testCase "reading produces decrypted results" $ do
-        let testLines = parseConfCrypt "test file" testFile
+        let filePath = "testFile"
+        lines <- T.readFile filePath
```
Comment (ChrisCoffey) [inline test/ConfCrypt/Commands/Tests.hs]: Does this no longer work using the default `testFile`?

File: testFile [hunk 9]
```diff
@@ -0,0 +1,4 @@
+Test : String
```
Comment (ChrisCoffey) [inline testFile]: Why move this to a separate file?

**Comment by ChrisCoffey** [comment]
@jhenligne thank you for the PR! The work is pretty close thus far, it just needs a little bit of polish before its ready to merge.

File: src/ConfCrypt/Commands.hs (hunk 1 above)
[the remark this answers is not in the document]
Comment (jhenligne) [reply src/ConfCrypt/Commands.hs]: With @ChrisCoffey explanations I removed last references to `MonadIO`.

File: app/ConfCrypt/CLI/Engine.hs (hunk 2 above)
Comment (jhenligne) [reply app/ConfCrypt/CLI/Engine.hs]: OK. I have removed it for now and kept it for another PR.

File: src/ConfCrypt/Commands.hs (hunk 3 above)
Comment (jhenligne) [reply src/ConfCrypt/Commands.hs]: That's right! Corrected in last commit.

File: src/ConfCrypt/Validation.hs (hunk 4 above)
Comment (jhenligne) [reply src/ConfCrypt/Validation.hs]: Errors are now accumulated in last commit.

File: src/ConfCrypt/Validation.hs (hunk 5 above)
Comment (jhenligne) [reply src/ConfCrypt/Validation.hs]: Errors are now accumulated in last commit.

File: src/ConfCrypt/Validation.hs (hunk 6 above)
Comment (jhenligne) [reply src/ConfCrypt/Validation.hs]: Done in last commit.

File: test/ConfCrypt/Commands/Tests.hs (hunk 7 above)
Comment (jhenligne) [reply test/ConfCrypt/Commands/Tests.hs]: Test restored in last commit.

File: test/ConfCrypt/Commands/Tests.hs (hunk 8 above)
Comment (jhenligne) [reply test/ConfCrypt/Commands/Tests.hs]: Yes, in my first implementation it was looking for a real file.
With last commit it has been restored.

File: testFile (hunk 9 above)
Comment (jhenligne) [reply testFile]: See previous comment.

File: src/ConfCrypt/Validation.hs [hunk 10]
```diff
@@ -22,71 +22,69 @@ import ConfCrypt.Types
...
-    -> m ()
+    -> m [T.Text]
 parameterTypesMatchSchema key ConfCryptFile {parameters} =
-    traverse_ decryptAndCompare parameters
+    traverse decryptAndCompare parameters
     where
         decryptAndCompare Parameter {paramName, paramValue, paramType} =
             catchError (runRule paramType paramName =<< decryptValue key paramValue)
-                       (const $ tell ["Error: Could not decrypt " <> paramName])
+                       (pure $ pure ("Error: Could not decrypt " <> paramName))
```
[the remark this answers is not in the document]
Comment (jhenligne) [reply src/ConfCrypt/Validation.hs]: I already tried but got a mismatch between `m` and `[]`.

File: test/ConfCrypt/Commands/Tests.hs [hunk 11]
```diff
@@ -11,10 +11,11 @@ import ConfCrypt.Encryption (unpackPrivateRSAKey, project, TextKey(..))
 import ConfCrypt.Common
 
 import Conduit (runResourceT)
+import Control.Monad (join)
 import Control.Monad.Identity (runIdentity)
 import Control.Monad.Reader (runReaderT)
 import Control.Monad.Except (runExcept, runExceptT)
-import Control.Monad.Writer (execWriter, execWriterT)
+--import Control.Monad.Writer (execWriter, execWriterT)
```
[the remark this answers is not in the document]
Comment (jhenligne) [reply test/ConfCrypt/Commands/Tests.hs]: Done.

**Comment by jhenligne** [comment]
@ChrisCoffey Thanks for your help and encouragements.
I don't know why it doesn't work with `pure [""]`, I tried many combinations but only find `pure $ pure ""` to work. I admit I don't find it much pleasing too.

File: src/ConfCrypt/Validation.hs (hunk 10 above)
[the remark this answers is not in the document]
Comment (fsoikin) [reply src/ConfCrypt/Validation.hs]: The second argument of `catchError` is supposed to be a function `e -> m a`, where `a` is instantiated here as `T.Text`.
There is an instance of `Applicative` for functions, where `pure` is implemented as `const`, so this works:

```
f :: Int -> String
f = pure "foo"

f 42 == "foo"
f 5 == "foo"
```

So in this case `pure $ pure "Error ..."` is equivalent to `const $ pure "Error ..."` or `\_ -> pure "Error ..."`.
The first `pure` comes from the `Applicative (a -> b)` instance, and the second pure comes from `m`.

The bottom line is that the first `pure` is technically ok (since it stands for `const`), but confusing, as this comment thread clearly demonstrates.

[the remark this answers is not in the document]
Comment (jhenligne) [reply src/ConfCrypt/Validation.hs]: Thanks, it helps. Commited.
